### PR TITLE
feat(metrics): Add derived metrics to meta endpoint [INGEST-924] [INGEST-1066]

### PR DESCRIPTION
**Pull request by ahmedetefy**
This PR:
- Adds support for derived metrics in Meta Endpoint 
- Redefines `MetricType` to be `counter, set, distribution, numeric`
- Adds a `generate_operations` method on `DerivedMetric` abstract base class

### CODE REVIEW

File: src/sentry/snuba/metrics/datasource.py [hunk 1]
```diff
@@ -68,17 +90,21 @@ def get_metrics(projects: Sequence[Project]) -> Sequence[MetricMeta]:
             org_id=projects[0].organization_id,
         )
     )
+    metric_names_meta_gen = (
```
Comment (untitaker) [inline src/sentry/snuba/metrics/datasource.py:93]: I quite dislike how this function is written... just write a for-loop that appends to a list! You apparently need a sorted list later anyway.

File: src/sentry/snuba/metrics/fields/base.py [hunk 2]
```diff
@@ -245,8 +247,18 @@ def __recursively_get_all_entities_in_derived_metric_dependency_tree(
...
         entities = set()
         derived_metric = DERIVED_METRICS[derived_metric_name]
+
+        try:
+            # Stop the recursion if the entity for this derived metric was computed previously
+            entity = derived_metric.get_entity()
+            if derived_metric.get_entity() is not None:
+                return {entity}
+        except DerivedMetricParseException:
+            ...
```
Comment (untitaker) [inline src/sentry/snuba/metrics/fields/base.py:260]: It seems we do this throughout the file... can we just use `pass`?

**Review by untitaker — APPROVED** [review]
the changes are fine but looking at the file y'all are being too clever for your own good

File: src/sentry/snuba/metrics/fields/base.py (hunk 2 above)
Comment (ahmedetefy) [reply src/sentry/snuba/metrics/fields/base.py:260]: why?

**Comment by ahmedetefy** [comment]
> the changes are fine but looking at the file y'all are being too clever for your own good

Can you elaborate on that? @untitaker

File: src/sentry/snuba/metrics/fields/base.py (hunk 2 above)
Comment (untitaker) [reply src/sentry/snuba/metrics/fields/base.py:260]: I don't think this is a common pattern to use, but I will follow up with you on slack

**Review by jjbayer — APPROVED** [review]

**Pull request merged**